daemon/prompter.c

```diff
@@ -495,15 +527,28 @@ prompter_eval_state_now(prompter_t *self)
     case PROMPTER_STATE_PROMPT:
         if( prompter_current_invocation(self) ) {
             /* We have a pending call */
+            if( self->prm_prompt )
+                prompter_set_state(self, PROMPTER_STATE_WAIT);
         }
         else if( !prompter_next_invocation(self) ) {
             prompter_set_state(self, PROMPTER_STATE_DISCONNECT);
         }
         else if( !prompter_prompt_invocation(self) ) {
-            prompter_fail_invocation(self);
+            prompter_set_state(self, PROMPTER_STATE_PROMPTING_FAILURE);
         }
         break;
 
+    case PROMPTER_STATE_WAIT:
+        if( !prompter_current_invocation(self) )
+            /* We resolved the pending call */
+            prompter_set_state(self, PROMPTER_STATE_PROMPT);
+        break;
+
+    case PROMPTER_STATE_CANCEL:
+        if( prompter_cancel_invocation(self) )
+            prompter_set_state(self, PROMPTER_STATE_PROMPT);
```

Comment:
Hmm..? This is in eval_state and we need to evaluate whether we may cancel the prompt whenever one of the conditions to do it is updated. It can't be done if we don't have prompt object path but when that is received the state is evaluated again.

Reply:
I changed this part a little by adding the check here instead of looking at prompter_cancel_invocation() reply value. (Though, now it does if (prm_prompt) check twice...)

Reply:
To clarify: Originally I assumed that 'cancel' would be analogous with 'connect' / 'disconnect' i.e. state has exactly one thing to do, it is initiated on entry to state (implicitly done only once), and state evaluation just detects when the action is finished (done in such manner that excess evaluation rounds have no effect) and we can move on.

Then on closer look: there is no on-entry action, but actions are taken in evaluator, and it looks like logic wise it is possible to end up in a situation where stm gets stuck at cancel state. As it seems that cancellation does not involve any async wait states  = no need to linger -> afaict it all could / should be handled on entry to state: just cancel what there is to cancel and then unconditionally transfer onward.

Reply:
The prompt can get cancelled from daemon point of view before `newPermissionPrompt` returns and at that point we can't do `cancel` call to windowprompt. It can be done only after `prompter_prompt_invocation_cb` has got the prompt object path and that's why I put it in evaluator.

Reply:
I'm starting to think: If cancelling is effectively just modifying what happens within WAIT state, It could be "cancel-when-it-becomes-possible" boolean prop that changes behavior on the way to prompter_prompt_invocation_cb() and within WAIT evaluator instead of being separate state with overlapping external dependencies.

Reply:
That makes the changes simpler, which hopefully limits risk for regressions. It should also work quite alright.